src/primitives/transaction.h

```diff
@@ -500,23 +585,33 @@ struct CMutableTransaction
         }
         nVersion = this->nVersion;
         if (fOverwintered) {
+            nVersion |= 1 << 31;
             READWRITE(nVersionGroupId);
         }
 
         bool isOverwinterV3 = fOverwintered &&
                               nVersionGroupId == OVERWINTER_VERSION_GROUP_ID &&
-                              nVersion == 3;
-        if (fOverwintered && !isOverwinterV3) {
+                              this->nVersion == 3;
+        bool isSaplingV4 =
+            fOverwintered &&
+            nVersionGroupId == SAPLING_VERSION_GROUP_ID &&
+            this->nVersion == SAPLING_TX_VERSION;
+        if (fOverwintered && !(isOverwinterV3 || isSaplingV4)) {
             throw std::ios_base::failure("Unknown transaction format");
         }
 
         READWRITE(vin);
         READWRITE(vout);
         READWRITE(nLockTime);
-        if (isOverwinterV3) {
+        if (isOverwinterV3 || isSaplingV4) {
             READWRITE(nExpiryHeight);
         }
-        if (nVersion >= 2) {
+        if (isSaplingV4) {
+            READWRITE(valueBalance);
+            READWRITE(vShieldedSpend);
+            READWRITE(vShieldedOutput);
+        }
+        if (this->nVersion >= 2) {
```

Comment:
We should also update `CTransaction::ToString()` below this to accomodate all the changes.